### PR TITLE
Sklearn api compliance

### PR DESCRIPTION
**Pull request by SaritNike**
In the branch [feat/sklearn-api-convention ](https://github.com/vc1492a/PyNomaly/tree/feat/sklearn-api-convention) we have realized that quite a lot was actually still missing to be fully compliant with the sklearn API. This is why we have decided to add and test these new fixes first in a separate branch.
Now that the check-api-compliance test is successfully passing we are ready to merge it all into the [feat/sklearn-api-convention ](https://github.com/vc1492a/PyNomaly/tree/feat/sklearn-api-convention) branch

### CODE REVIEW

File: PyNomaly/loop.py [hunk 1]
```diff
@@ -110,166 +138,317 @@ def __init__(
...
+        if y is not None and distance_matrix is None:
+            y_array = np.asarray(y)
+            if y_array.ndim == 2:
+                distance_matrix = y
+                y = None
+                warnings.warn(
+                    "Passing 'distance_matrix' as the second positional "
+                    "argument is deprecated. Use keyword arguments instead.",
+                    FutureWarning,
+                )
```
Comment (vc1492a) [inline PyNomaly/loop.py]: **Hazard: a 2-d `y` is silently hijacked as `distance_matrix`.**

A scikit-learn user calling `fit(X, y)` with labels shaped `(n, 1)` (e.g. from `reshape(-1, 1)` or a single-column DataFrame) hits this branch: `y` becomes `distance_matrix`, validation then warns "only one of data or distance matrix" and `fit()` returns an **unfitted** model (`is_fit_=False`) with only warnings. Verified locally:

```python
clf = LoOP(n_neighbors=5).fit(X, np.ones((30, 1)))  # is_fit_ == False, no exception
```

Suggest tightening the guard so the legacy fallback only applies when it can't be a label vector, e.g.:

```python
if y is not None and distance_matrix is None and data is None:
    y_array = np.asarray(y)
    if y_array.ndim == 2 and y_array.shape[1] > 1:
        ...
```

(`data is None` because the legacy positional pattern was `fit(distance_matrix=...)` without raw data; a distance matrix with a single neighbor column is rare enough that requiring keywords there is reasonable.)

File: PyNomaly/loop.py [hunk 2]
```diff
@@ -110,166 +138,317 @@ def __init__(
...
+        )
+        self.neighbor_matrix_ = (
+            neighbor_matrix if neighbor_matrix is not None else self.neighbor_matrix
+        )
+        self.cluster_labels_ = (
+            cluster_labels if cluster_labels is not None else self.cluster_labels
+        )
 
         if self._validate_inputs() is False:
             return self
```
Comment (vc1492a) [inline PyNomaly/loop.py:253]: **`fit()` silently returns an unfitted model on invalid input.**

When validation fails we warn and `return self` with `is_fit_ = False`. Downstream `predict()`/`decision_function()` do eventually raise `NotFittedError`, so nothing breaks silently — but sklearn convention is to raise `ValueError` from `fit()` itself. Fine to keep for 1.0.0 backward compatibility (the old API behaved this way), but worth revisiting in the follow-up deprecation PR: warn-and-return today, raise in a future version.

File: PyNomaly/loop.py [hunk 3]
```diff
@@ -110,166 +138,317 @@ def __init__(
...
+            self.n_features_in_ = 1
+
+        # sklearn outlier API: decision_function = score_samples - offset_
+        # score_samples = -LoOP probability (higher is more normal)
+        # offset_ = -0.5 so decision >= 0 iff probability <= 0.5
+        self.offset_ = -0.5
+        self.is_fit_ = True
         return self
 
     def stream(self, x: np.ndarray) -> np.ndarray:
```
Comment (vc1492a) [inline PyNomaly/loop.py]: Nit: the return annotation still says `np.ndarray`, but `stream()` now returns a Python `float` (see the `float(np.asarray(loop_score, ...).ravel()[0])` at the end). Should be `-> float`. Also worth a line in the changelog since callers doing `.item()` or array ops on the result will see a type change.

File: docs/api.md [hunk 4]
```diff
@@ -22,7 +22,7 @@ clf = loop.LocalOutlierProbability(
 
 | Parameter | Type | Default | Description |
 |---|---|---|---|
-| `data` | `np.ndarray` or `pd.DataFrame` | `None` | Input data as a 2D array with shape (n_observations, n_features). Mutually exclusive with `distance_matrix`. |
+| `data` | `np.ndarray`, `pd.DataFrame`, or `scipy.sparse` matrix | `None` | Input data as a 2D array with shape (n_observations, n_features). Sparse matrices are accepted and densified internally (requires scipy). Mutually exclusive with `distance_matrix`. |
```
Comment (vc1492a) [inline docs/api.md:25]: Docs in this file still reference the legacy non-underscore attributes (`clf.local_outlier_probabilities` on line 50, the attribute table at lines 86/90). They still work as read-only property aliases, so nothing is wrong — but since fitted state now canonically lives on `local_outlier_probabilities_`, `distance_matrix_`, etc., the docs should show the underscore names as primary and mention the aliases as deprecated. Good candidate for the follow-up PR that adds warnings on legacy attribute access (note: `distance_matrix` after fit is the *init input*, not the computed matrix — that one has no alias by design and is the main read-side breaking change to document).

**Review by vc1492a — COMMENTED** [review]
Deep review done (manual + AI-assisted). Overall this is in good shape and ready to merge into the feature branch once the `fit(X, y)` comment below is addressed or consciously deferred.

**Verified locally against head `fa903ec`:**
- Full test suite: 46 passed, 4 skipped
- `check_estimator(LoOP())`: all checks pass (one documented xfail: `check_fit1d`, since LoOP intentionally accepts 1-d input and reshapes to `(-1, 1)`)
- Param immutability: `get_params()` identical before/after `fit()` (effective values live on `n_neighbors_`, `_fit_n_jobs`, etc.)
- Re-fit across input modes (raw data → distance matrix) produces correct, consistent scores
- Sparse CSR/CSC: fit/predict/stream work, dense/sparse scores identical, clear `ImportError` without scipy
- `predict`/`decision_function`/`score_samples` are mutually consistent (`decision = scores - offset_`, labels from `decision >= 0`)
- Distance-matrix mode: `n_features_in_ = 1`, predict works on single-feature input

**Findings (line comments):**
1. `fit(X, y)` with a `(n, 1)` label array silently becomes `distance_matrix` and returns an unfitted model — the one thing I'd fix before merge
2. `fit()` warns-and-returns on invalid input instead of raising `ValueError` (acceptable for BC, revisit in the deprecation-warnings PR)
3. `stream()` return annotation says `np.ndarray` but now returns `float`
4. `docs/api.md` still shows legacy non-underscore attribute names as primary

On the read-side breaking change we discussed: confirmed `distance_matrix` after fit is still the init input (often `None`), and the computed matrix is only on `distance_matrix_`. That's correct per sklearn's config/state firewall — it just needs to be loud in the changelog and the follow-up warnings PR.

No automation will merge this — approving only records the review; merge stays manual.

File: PyNomaly/loop.py (hunk 3 above)
Comment (vc1492a) [reply PyNomaly/loop.py]: Resolved in 80c8f44: `stream()` annotation is now `-> float`, API docs updated to match, and the 1.0.0 changelog notes that `stream()` returns a Python `float` instead of a NumPy array.

File: PyNomaly/loop.py (hunk 1 above)
Comment (vc1492a) [reply PyNomaly/loop.py]: Resolved in 9f74189. The legacy fallback now only triggers when no raw data is given (`data is None`) **and** `y` has more than one column, so sklearn-style `fit(X, y)` with `(n, 1)` labels fits normally. Verified the original repro now yields `is_fit_ == True`, the legacy positional `fit(None, distance_matrix)` path still works with its `FutureWarning`, and added a regression test (`test_fit_with_2d_label_y_is_ignored`). Full suite: 47 passed.

**Comment by vc1492a** [comment]
@SaritNike finished my review and addressed two small issues I identified with the help of Fable. I also took some time to update our documentation and example scripts to reflect the changes in this PR. 

Please take another look. If this looks good to you, let's go ahead and merge in and address the deprecation warnings in a follow up.

File: PyNomaly/loop.py (hunk 2 above)
Comment (SaritNike) [reply PyNomaly/loop.py:253]: good catch

**Comment by SaritNike** [comment]
I took another round and reviewed it. we can merge it into the feature branch

**Comment by SaritNike** [comment]
I see that you have added more commits. Feel free to merge when you are ready
